libazureinit/src/provision/ssh.rs

```diff
@@ -29,10 +29,12 @@ pub(crate) fn provision_ssh(
     // the permissions are correct.
     std::fs::set_permissions(&ssh_dir, Permissions::from_mode(0o700))?;
 
-    let authorized_keys_path = match get_authorized_keys_path_from_sshd()? {
-        Some(path) => user.dir.join(path),
-        None => user.dir.join(".ssh/authorized_keys"),
-    };
+    let authorized_keys_path_string = authorized_keys_path_string
+        .unwrap_or_else(|| {
+            get_authorized_keys_path_from_sshd(get_sshd_g_output)
+                .unwrap_or_else(|| ".ssh/authorized_keys".to_string())
+        });
```

Comment:
Two tweaks you could do here (which are equivalent to what you've got):

```Rust
    let authorized_keys_path_string = authorized_keys_path_string
        .or_else(|| {
            get_authorized_keys_path_from_sshd(|| {
                Command::new("sshd").arg("-G").output()
            })
        })
        .unwrap_or_else(|| ".ssh/authorized_keys".to_string());
```
Here I used [or_else](https://doc.rust-lang.org/std/option/enum.Option.html#method.or_else) instead of nesting the `unwrap_or_else`. It's a slightly nicer way to chain `Option`s you're trying to pick from.

The other thing I did was just use a lambda for `fn get_sshd_g_output` which is only used once. I assume things are organized like this to make mocking out the command easy.

FWIW sshd does let you specify a config and, assuming it can read the config with user permissions, doesn't need elevated privileges. So you could accept a `Command` and then in the tests do this kind of thing:
```Rust
        let mut sshd_config = NamedTempFile::new().unwrap();
        sshd_config.write_all("authorizedkeysfile .ssh/test_authorized_keys".as_bytes()).unwrap();
        let mut command = Command::new("sshd");
        command.arg("-G").arg(format!("-f {:?}", sshd_config.path()));

        let result = get_authorized_keys_path_from_sshd(command);
```

But it does mean the tests depend on running on a host new enough to have `-G` (which I _think_ is reasonable, but maybe not).

Anyway, up to you if you go with this approach.

Reply:
> I left a lot of little tips-n-tricks notes, but it looks in pretty good shape to me. In case you want to get a sense of what the logs look like, I left some notes in #135 on how to get logs out of tests.

Great feedback, thanks!  Appreciate all the extra help and explanations to improve my non-existent Rust skills :D

Reply:
fixed!